docs/langref/relay_expr.rst

```diff
@@ -0,0 +1,396 @@
+and documentation.
+
+Local Variable
+~~~~~~~~~~~~~~~~~
+
+Local identifiers are prefixed by the :code:`%` sigil,
+such as ":code:`%local`". A local identifier always references
+a function argument or a variable bound in a :code:`let` expression.
+A local variable will be scoped to the function where it
+appears or the :code:`let` expression where it is bound, respectively.
+
+Suppose the local variable :code:`%a` has been defined in a scope
+and within that scope, a :code:`let` expression binding to a variable
+:code:`%a` appears. This is permitted, as in most functional languages.
+In the scope of the :code:`let` expression (the inner scope),
+the name :code:`%a` is "shadowed," meaning all references to
+:code:`%a` in the inner scope refer to the later defintion, while
+references to :code:`%a` in the outer scope continue to refer to
+the previous one.
+
+(Note that in Relay's implemetnation, each definition of a local variable
+creates a new :py:class:`~tvm.relay.expr.Var`, so a shadowed local variable,
+despite having the same name as one in an outer scope, will be a different
+object. This allows for comparing local variables by pointer identity with the
+knowledge that the same local variable object corresponds to a different binding site.)
+
+See :py:class:`~tvm.relay.expr.Var` for its implementation
+and documentation.
+
+Global Functions
+================
+
+A function definition consists of a name, arguments, return type,
+type parameters, and any applicable type relations.
+A function's return type and parameter types may be omitted;
+Relay will attempt to infer the most general types where types
+are omitted.
+
+Functions in Relay act similarly to procedures or functions in
+other programming language and serve to generalize the concept
+of a named subgraph.
+
+Functions defined in the manner described in this subsection are
+of global scope; anonymous functions will be discussed later, though
+their mechanics are nearly identical. Note that global functions may be
+recursive; that is, within the function's body, the function's
+identifier refers back to the function unless it is shadowed in a :code:`let`
+expression.
+
+A definition minimally consists of an identifier :code:`@id`, an empty set of
+parameters, and a body expression (:py:class:`~tvm.relay.expr.Expr`)
+contained by curly braces.
+
+.. code-block:: python
+
+    def @id() { body }
+
+A definition may contain any number of parameters. For example, a
+simple function that invokes the `add` operator:
+
+.. code-block:: python
+
+    def @plus(%x, %y) { add(%x, %y) }
+
+Notice that within the function's body, the parameters are local
+variables, just like those bound in a :code:`let` expression.
+
+One may also annotate explicit types on definitions.
+For example, we can restrict the above definition to only work
+on certain types:
+
+.. code-block:: python
+
+    def @plus(%x: Tensor<(10, 10), float32>, %y: Tensor<(10, 10), float32>)
```

Comment:
syntax need to be updated to the new proposed syntax

Reply:
Which part needs to be changed? I see that the examples @joshpoll posted use square brackets on tensor types and `fn` instead of `fun` for function literals. I will change those, but are there other inconsistencies?

Reply:
Ah and no brackets around if-else blocks

Reply:
I recalled it was ```Tensor[(10, 10), float32]```?

Reply:
sorry i was commenting on stale content, but yes we need brackets around if-else

Reply:
I'll put the brackets back then (only two places). The actual ANTLR AST for if-else exprs doesn't appear to require brackets, so we'll have to sync those specs

Reply:
The ANTLR grammar _does_ require brackets. It uses a body expression. `test_ir_parser.py` contains examples for each language construct.

Reply:
this example is fine except for the tensor types